### PR TITLE
{devel}[system] ReFrame v3.5.3

### PR DESCRIPTION
**Pull request by teojgo**

### CODE REVIEW

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3101.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/35b60f404d4da543783dc81bcad70238 for a full test report.

File: easybuild/easyconfigs/r/ReFrame/ReFrame-3.5.3.eb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+allow_system_deps = [('Python', SYS_PYTHON_VERSION)]
+
+req_py_majver = 3
+req_py_minver = 6
+
+use_pip = True
+
+exts_list = [
+    ('pip', '21.0.1', {
```
Comment (ocaisa) [inline easybuild/easyconfigs/r/ReFrame/ReFrame-3.5.3.eb:19]: This is not actually installed properly the last time I tested this, if pip is not available in calling Python the build will fail.

Comment (boegel) [reply easybuild/easyconfigs/r/ReFrame/ReFrame-3.5.3.eb:19]: I see `software/ReFrame/3.5.3/bin/pip` installed?

Comment (ocaisa) [reply easybuild/easyconfigs/r/ReFrame/ReFrame-3.5.3.eb:19]: Yes, but if one of the subsequent modules tries to use it for installation, they fail.

Comment (ocaisa) [reply easybuild/easyconfigs/r/ReFrame/ReFrame-3.5.3.eb:19]: You know Python better than me, but the pip gets installed with an eggdir that doesn't get picked up with `use_pip`, so `bin/pip` is there but not the module.

Comment (ocaisa) [reply easybuild/easyconfigs/r/ReFrame/ReFrame-3.5.3.eb:19]: See https://github.com/easybuilders/easybuild-easyconfigs/issues/12326

Comment (boegel) [reply easybuild/easyconfigs/r/ReFrame/ReFrame-3.5.3.eb:19]: The `pip --version` check that is done right before installing the first extension with `pip install` works fine though:

```
== 2021-04-22 11:38:51,955 run.py:571 DEBUG cmd "pip --version" exited with exit code 0 and output:
pip 21.0.1 from /user/gent/400/vsc40023/eb_arcaninescratch/CO7/skylake-ib/software/ReFrame/3.5.3/lib/python3.6/site-packages/pip-21.0.1-py3.6.egg/pip (python 3.6)
```

So I don't think the `.egg` directory doesn't seem to be a problem.

This is with Python 3.6 though, maybe it's only a problem with more recent Python versions?

**Comment by akesandgren** [comment]
Test report by @akesandgren
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
b-cn0235.hpc2n.umu.se - Linux Ubuntu 20.04, x86_64, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 3.8.5
See https://gist.github.com/d82f0b59ce324ab23830e59637fa1c5a for a full test report.

**Comment by akesandgren** [comment]
Also see my test report.

**Comment by akesandgren** [comment]
Looks like pip needs setuptools to be there first.

**Comment by akesandgren** [comment]
Tried moving setuptools before pip (with use_pip: False) but that fails, then tried:
```
exts_list = [
    ('setuptools', '53.0.0', {
        'source_tmpl': 'setuptools-%(version)s.tar.gz',
        'use_pip': False,
        'checksums': ['1b18ef17d74ba97ac9c0e4b4265f123f07a8ae85d9cd093949fa056d3eeeead5'],
    }),
    ('pip', '21.0.1', {
        'use_pip': False,
        'prebuildopts': ("export PYTHONPATH=%(installdir)s/lib/python3.8/site-packages:$PYTHONPATH && "),
        'checksums': ['99bbde183ec5ec037318e774b0d8ae0a64352fe53b2c7fd630be1d07e94f41e5'],
    }),
```
But it still fails with:
```
# output for command: export PYTHONPATH=/home/a/ake/Public/easybuild/software/ReFrame/3.5.3/lib/python3.8/site-packages:$PYTHONPATH &&  /usr/bin/python3 setup.py build

Traceback (most recent call last):
  File "setup.py", line 7, in <module>
    from setuptools import find_packages, setup
ModuleNotFoundError: No module named 'setuptools'
```

**Comment by akesandgren** [comment]
This very ugly hack gives me an installed ReFrame, still broken due to missing deps but ...
```
exts_list = [
    ('pip', '21.0.1', {
        'use_pip': False,
        'prebuildopts': ("curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py && python3 ./get-pip.py --prefix %(installdir)s && "),
        'checksums': ['99bbde183ec5ec037318e774b0d8ae0a64352fe53b2c7fd630be1d07e94f41e5'],
    }),
    ('setuptools', '53.0.0', {
        'source_tmpl': 'setuptools-%(version)s.tar.gz',
        'checksums': ['1b18ef17d74ba97ac9c0e4b4265f123f07a8ae85d9cd093949fa056d3eeeead5'],
    }),
    ('reframe', version, {
        # Bump argcomplete to 1.12.3 ensuring compatibility with importlib-metadata >= 4.0
        'preinstallopts': ("sed -i 's/argcomplete==1.12.2/argcomplete==1.12.3/g' requirements.txt && "
                           "./bootstrap.sh +docs +pygelf && cp -r external %(installdir)s && "),
        'source_tmpl': 'v%(version)s.tar.gz',
        'source_urls': ['https://github.com/eth-cscs/reframe/archive/'],
        'checksums': ['5e7f8b93cd672dcb702c657ec2f595d34d3335b1d16484a596a083b5ef81d7ec'],
    }),
]
```

**Comment by akesandgren** [comment]
There is also a dep on pycairo, from pygobject 3.36.0

**Comment by teojgo** [comment]
> There is also a dep on pycairo, from pygobject 3.36.0

Does this come from ReFrame or from setuptools?

**Comment by akesandgren** [comment]
Don't know. Probably not ReFrame since I can't find a dep for that in there.

**Comment by casparvl** [comment]
I was struggling with installing `ReFrame-3.6.2` (EasyConfig based on this PR) too. First, it complained that
```
reframe-hpc 3.6.2 requires argcomplete, which is not installed.
reframe-hpc 3.6.2 requires lxml, which is not installed.
reframe-hpc 3.6.2 requires pyyaml, which is not installed.
```
Which are indeed all in the requirements file, but not in the EasyConfig. Adding them in the EasyConfig is an option, but then my `lxml` failed because our system python doesn't have development headers (`Python.h`) installed.

I finally made it work using `use_pip_for_deps` and `download_dep_fail`:
```
    ('reframe', version, {
        'preinstallopts': """sed -i "s/\(importlib.*\)==.*/,==3.10.0; python_version < '3.8'/g" requirements.txt && ./bootstrap.sh +docs +pygelf && cp -r external %(installdir)s && """,
        'source_tmpl': 'v%(version)s.tar.gz',
        'source_urls': ['https://github.com/eth-cscs/reframe/archive/'],
        'use_pip_for_deps': True,
        'download_dep_fail': False,
        'checksums': ['b4725e434f4cd721ca825a56a652e67aa77e7af5ed7ca00f281747585827060d'],
    }),
    ```

Now, I know we dislike binary installs in `EasyBuild` and prefer to build from sources. However, considering that

- `ReFrame` is (necessarily) at the `SYSTEM` level
- We don't have control over the Python environment offered for SYSTEM(e.g. if development headers are present)
- `ReFrame` itself is not performance critical

Is it an idea to do this for `ReFrame` EasyConfigs from now on? My guess would be that this gives the fewest issues overall for EB users... What do you guys think?

**Comment by teojgo** [comment]
> I was struggling with installing `ReFrame-3.6.2` (EasyConfig based on this PR) too. First, it complained that
> 
> ```
> reframe-hpc 3.6.2 requires argcomplete, which is not installed.
> reframe-hpc 3.6.2 requires lxml, which is not installed.
> reframe-hpc 3.6.2 requires pyyaml, which is not installed.
> ```
> 
> Which are indeed all in the requirements file, but not in the EasyConfig. Adding them in the EasyConfig is an option, but then my `lxml` failed because our system python doesn't have development headers (`Python.h`) installed.
> 
> I finally made it work using `use_pip_for_deps` and `download_dep_fail`:
> 
> ```
>     ('reframe', version, {
>         'preinstallopts': """sed -i "s/\(importlib.*\)==.*/,==3.10.0; python_version < '3.8'/g" requirements.txt && ./bootstrap.sh +docs +pygelf && cp -r external %(installdir)s && """,
>         'source_tmpl': 'v%(version)s.tar.gz',
>         'source_urls': ['https://github.com/eth-cscs/reframe/archive/'],
>         'use_pip_for_deps': True,
>         'download_dep_fail': False,
>         'checksums': ['b4725e434f4cd721ca825a56a652e67aa77e7af5ed7ca00f281747585827060d'],
>     }),
> ```
> 
> Now, I know we dislike binary installs in `EasyBuild` and prefer to build from sources. However, considering that
> 
> * `ReFrame` is (necessarily) at the `SYSTEM` level
> * We don't have control over the Python environment offered for `SYSTEM` (e.g. if development headers are present)
> * `ReFrame` itself is not performance critical
> 
> Is it an idea to do this for `ReFrame` EasyConfigs from now on? My guess would be that this gives the fewest issues overall for EB users... What do you guys think?

Let me try myself and come back to you. In general, the installation relies on the bootstrap script of reframe which downloads all the required dependencies.

**Comment by casparvl** [comment]
Interesting, so we shouldn't _have_ to add those requirements to the `EasyConfig`. Not sure why in my case the `pip check` didn't pick up on them then...

**Comment by teojgo** [comment]
> Interesting, so we shouldn't _have_ to add those requirements to the `EasyConfig`. Not sure why in my case the `pip check` didn't pick up on them then...

Can you try this recipe? I will make a PR and close this if it's fine.

**Comment by casparvl** [comment]
@teojgo : Just tried it, and the installation completed without issues :) I guess it's the export of the `PYTHONPATH` to include the external deps that fixed it... If you make a PR, let me know: I can upload a test report for it.

**Comment by boegel** [comment]
closing this since #13096 got merged

**Pull request closed without merging**